File: src/lib/server/llm/openrouter.ts
import type { CorrectionResult, LLMProvider } from './types';

export class OpenRouterProvider implements LLMProvider {
	private apiKey: string;
	private model: string;

	constructor(apiKey: string, model = 'openai/gpt-4o-mini') {
		this.apiKey = apiKey;
		this.model = model;
	}

	async correctText(text: string): Promise<CorrectionResult> {
		const controller = new AbortController();
		const timeout = setTimeout(() => controller.abort(), 30_000);

		try {
			const res = await fetch('https://openrouter.ai/api/v1/chat/completions', {
				method: 'POST',
				headers: {
					Authorization: `Bearer ${this.apiKey}`,
					'Content-Type': 'application/json'
				},
				body: JSON.stringify({
					model: this.model,
					messages: [
						{
							role: 'system',
							content: 'Return ONLY the corrected text, no commentary.'
						},
						{ role: 'user', content: text }
					]
				}),
				signal: controller.signal
			});

			if (!res.ok) {
				const body = await res.text();
				throw new Error(`OpenRouter API error ${res.status}: ${body}`);
			}

			const data = await res.json();
			const correctedText = data.choices?.[0]?.message?.content?.trim() ?? text;
			return { correctedText };
		} finally {
			clearTimeout(timeout);
		}
	}
}

File: src-tauri/src/inference/openrouter.rs
use async_trait::async_trait;
use serde::Deserialize;

use super::{LLMProvider, SYSTEM_PROMPT};

#[derive(Deserialize)]
struct OpenRouterResponse {
    choices: Vec<Choice>,
}

#[derive(Deserialize)]
struct Choice {
    message: Message,
}

#[derive(Deserialize)]
struct Message {
    content: String,
}

pub struct OpenRouterProvider {
    api_key: String,
}

impl OpenRouterProvider {
    pub fn new(api_key: String) -> Self {
        Self { api_key }
    }
}

#[async_trait]
impl LLMProvider for OpenRouterProvider {
    async fn correct_text(&self, text: &str) -> Result<String, String> {
        let client = reqwest::Client::new();
        let response = client
            .post("https://openrouter.ai/api/v1/chat/completions")
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .timeout(std::time::Duration::from_secs(30))
            .json(&serde_json::json!({
                "model": "openai/gpt-4o-mini",
                "messages": [
                    {
                        "role": "system",
                        "content": SYSTEM_PROMPT
                    },
                    {
                        "role": "user",
                        "content": text
                    }
                ]
            }))
            .send()
            .await
            .map_err(|e| format!("Request failed: {}", e))?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().await.unwrap_or_default();
            return Err(format!("OpenRouter API error {}: {}", status, body));
        }

        let data: OpenRouterResponse = response
            .json()
            .await
            .map_err(|e| format!("Failed to parse response: {}", e))?;

        let corrected = data
            .choices
            .first()
            .map(|c| c.message.content.trim().to_string())
            .unwrap_or_else(|| text.to_string());

        Ok(corrected)
    }
}
